code/datums/actions/ability_actions.dm

```diff
@@ -98,11 +98,19 @@
 			carbon_owner.balloon_alert(carbon_owner, "Cannot while burrowed")
 		return FALSE
 
-	if(!(to_check_flags & ABILITY_USE_CLOSEDTURF) && isclosedturf(get_turf(carbon_owner)))
-		if(!silent)
-			//Not converted to balloon alert as xeno.dm's balloon alert is simultaneously called and will overlap.
-			to_chat(owner, span_warning("We can't do this while in a solid object!"))
-		return FALSE
+	if(!(to_check_flags & ABILITY_USE_SOLIDOBJECT))
+		var/turf/current_turf = get_turf(carbon_owner)
+		if(isclosedturf(current_turf))
+			if(!silent)
+				//Not converted to balloon alert as xeno.dm's balloon alert is simultaneously called and will overlap.
+				to_chat(owner, span_warning("We can't do this while in a solid object!"))
+			return FALSE
+		for(var/obj/machinery/door/found_door in current_turf.contents)
+			if(found_door.density)
+				if(!silent)
+					//Same as above.
+					to_chat(owner, span_warning("We can't do this while in a solid object!"))
+				return FALSE
```

Comment:
this seems pretty snowflake for doors, when presumably this behavior shouldn't be exclusive to just doors.
Would it be better to just check all objects, and see if they are both dense and opaque?

Reply:
Not sure how to deal with vehicles. If I checked for all objs instead of just doors, xenos who are ontop of vehicles won't be able to use their abilities.

Reply:
dense AND  opaque.